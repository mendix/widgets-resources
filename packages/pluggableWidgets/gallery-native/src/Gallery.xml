<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.gallery.Gallery" pluginWidget="true" supportedPlatform="Native" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Gallery</name>
    <description />
    <studioProCategory>Data containers</studioProCategory>
    <studioCategory>Data Containers</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/modules/gallery</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="datasource" type="datasource" isList="true">
                    <caption>Data source</caption>
                    <description />
                </property>
                <property key="content" type="widgets" dataSource="datasource" required="false">
                    <caption>Content placeholder</caption>
                    <description />
                </property>
                <property key="scrollDirection" type="enumeration" defaultValue="vertical">
                    <caption>Scroll direction</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="vertical">Vertical</enumerationValue>
                        <enumerationValue key="horizontal">Horizontal</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Columns">
                <property key="tabletColumns" type="integer" defaultValue="1">
                    <caption>Tablet columns</caption>
                    <description />
                </property>
                <property key="phoneColumns" type="integer" defaultValue="1">
                    <caption>Phone columns</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Items">
                <property key="pageSize" type="integer" defaultValue="20">
                    <caption>Page size</caption>
                    <description />
                </property>
                <property key="pagination" type="enumeration" defaultValue="virtualScrolling">
                    <caption>Pagination</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="virtualScrolling">Virtual scrolling</enumerationValue>
                        <enumerationValue key="buttons">Paging buttons</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="pagingPosition" type="enumeration" defaultValue="below">
                    <caption>Position of paging buttons</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="below">Below grid</enumerationValue>
                        <enumerationValue key="above">Above grid</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="pagingButtonText" type="string" defaultValue="Load more">
                    <caption>Title of paging button</caption>
                    <description />
                </property>
                <property key="showEmptyPlaceholder" type="enumeration" defaultValue="none">
                    <caption>Empty message</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="emptyPlaceholder" type="widgets" required="false">
                    <caption>Empty placeholder</caption>
                    <description />
                </property>
                <property key="itemClass" type="expression" required="false" dataSource="datasource">
                    <caption>Dynamic item class</caption>
                    <description />
                    <returnType type="String" />
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClick" type="action" required="false" dataSource="datasource">
                    <caption>On click action</caption>
                    <description />
                </property>
                <property key="pullDown" type="action" required="false">
                    <caption>Pull down</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
