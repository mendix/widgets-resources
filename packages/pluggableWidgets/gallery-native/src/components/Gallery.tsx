import { createElement, ReactElement, ReactNode } from "react";
import { Button, FlatList, Pressable, View } from "react-native";
import { ObjectItem } from "mendix";
import DeviceInfo from "react-native-device-info";
import { GalleryStyle } from "../ui/Styles";
import { PagingPositionEnum, ScrollDirectionEnum } from "../../typings/GalleryProps";

export interface GalleryProps<T extends ObjectItem> {
    emptyPlaceholderRenderer?: (renderWrapper: (children: ReactNode) => ReactElement) => ReactElement;
    hasMoreItems: boolean;
    isInfiniteLoad: boolean;
    itemRenderer: (
        renderWrapper: (children: ReactNode, className?: string, onClick?: () => void) => ReactElement,
        item: T
    ) => ReactElement;
    items: T[];
    loadMoreItems?: (computePage: (prevPage: number) => number) => void;
    name: string;
    numberOfItems?: number;
    page: number;
    pageSize: number;
    paginationPosition?: PagingPositionEnum;
    paging: boolean;
    phoneColumns: number;
    preview?: boolean;
    pullDown?: () => void;
    scrollDirection: ScrollDirectionEnum;
    style: GalleryStyle;
    tabIndex?: number;
    tabletColumns: number;
}

export const Gallery = <T extends ObjectItem>(props: GalleryProps<T>): ReactElement => {
    const listItemMargin = props.style.listItem?.margin || 0;
    const listItemPadding = props.style.listItem?.padding || 0;
    const columnSize = DeviceInfo.isTablet() ? props.tabletColumns : props.phoneColumns;
    const ListEmptyComponent =
        props.emptyPlaceholderRenderer && props.emptyPlaceholderRenderer(children => <View>{children}</View>);
    const onEndReached = (): any =>
        props.isInfiniteLoad && props.hasMoreItems && props.loadMoreItems && props.loadMoreItems(prev => prev + 1);
    const renderItem = (item: { item: T }): ReactElement =>
        item.item.id !== "_blank" ? (
            props.itemRenderer(
                (children, className, onClick) => (
                    <Pressable
                        testID={`${props.name}-list-item-${item.item.id}`}
                        style={[
                            props.style.listItem,
                            className ? props.style?.customClasses?.[className]?.listItem : [],
                            { flex: 1 / columnSize }
                        ]}
                        onPress={onClick}
                    >
                        {children}
                    </Pressable>
                ),
                item.item
            )
        ) : (
            <View
                style={{
                    flex: 1 / columnSize,
                    backgroundColor: "transparent",
                    padding: listItemPadding,
                    margin: listItemMargin
                }}
            />
        );
    const pagination =
        !props.isInfiniteLoad && props.paging && props.hasMoreItems ? (
            <View style={props.style.pagination}>
                <Button
                    testID={`${props.name}-pagination-button`}
                    onPress={() => props.hasMoreItems && props.loadMoreItems && props.loadMoreItems(prev => prev + 1)}
                    title={"Load more items..."}
                />
            </View>
        ) : undefined;

    const formatData = (dataList: T[], numColumns: number): T[] => {
        let totalLastRow = dataList.length % numColumns;
        while (totalLastRow !== 0 && totalLastRow !== numColumns) {
            const emptyData: any = { id: "_blank" };
            dataList.push(emptyData);
            totalLastRow++;
        }
        return dataList;
    };

    return (
        <View testID={props.name} style={props.style.container}>
            <FlatList
                testID={`${props.name}-list`}
                {...(props.scrollDirection === "vertical" && { onRefresh: props.pullDown })}
                data={formatData(props.items, columnSize)}
                horizontal={props.scrollDirection === "horizontal"}
                keyExtractor={item => item.id}
                ListEmptyComponent={ListEmptyComponent}
                ListFooterComponent={pagination}
                numColumns={columnSize}
                onEndReached={onEndReached}
                refreshing={false}
                renderItem={renderItem}
                style={props.style.listStyle}
            />
        </View>
    );
};
