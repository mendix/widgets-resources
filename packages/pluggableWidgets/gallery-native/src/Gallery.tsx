import { createElement, ReactElement, useCallback, useEffect } from "react";
import { all } from "deepmerge";
import { executeAction } from "@mendix/piw-utils-internal";
import { Gallery as GalleryComponent } from "./components/Gallery";
import { defaultGalleryStyle, GalleryStyle } from "./ui/Styles";
import { GalleryProps } from "../typings/GalleryProps";

export const Gallery = (props: GalleryProps<GalleryStyle>): ReactElement => {
    const styles = all<GalleryStyle>([defaultGalleryStyle, ...props.style]);
    const currentPage = props.datasource.limit / props.pageSize;

    useEffect(() => {
        if (props.datasource.limit === Number.POSITIVE_INFINITY) {
            props.datasource.setLimit(props.pageSize);
        }
    }, [props.datasource, props.pageSize]);

    const loadMoreItems = useCallback(() => {
        props.datasource.setLimit((currentPage + 1) * props.pageSize);
    }, [currentPage, props.datasource, props.pageSize]);

    const itemRenderer = useCallback(
        (renderWrapper, item) =>
            renderWrapper(
                props.content?.get(item),
                props.itemClass?.get(item)?.value,
                props.onClick ? () => executeAction(props.onClick?.get(item)) : undefined
            ),
        [props.content, props.itemClass, props.onClick]
    );

    const pullDown = useCallback(() => props.pullDown && executeAction(props.pullDown), [props.pullDown]);

    return (
        <GalleryComponent
            emptyPlaceholder={props.emptyPlaceholder}
            hasMoreItems={props.datasource.hasMoreItems ?? false}
            itemRenderer={itemRenderer}
            items={props.datasource.items ?? []}
            loadMoreItems={loadMoreItems}
            name={props.name}
            paginationPosition={props.pagingPosition}
            paging={props.pagination === "buttons"}
            pagingButtonText={props.pagingButtonText}
            phoneColumns={props.phoneColumns}
            pullDown={pullDown}
            pullDownIsExecuting={props.pullDown?.isExecuting ?? false}
            scrollDirection={props.scrollDirection}
            style={styles}
            tabletColumns={props.tabletColumns}
        />
    );
};
