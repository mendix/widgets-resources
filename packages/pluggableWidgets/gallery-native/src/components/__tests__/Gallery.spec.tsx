/**
 * @jest-environment jsdom
 */
import { createElement } from "react";
import { Text } from "react-native";
import { ObjectItem, GUID } from "mendix";
import { render, fireEvent, act } from "@testing-library/react-native";
import { Gallery, GalleryProps } from "../Gallery";

jest.mock("react-native-device-info", () => ({ isTablet: jest.fn().mockReturnValue(false) }));

const itemWrapperFunction =
    ({ onClick }: { onClick?: () => void }): GalleryProps<ObjectItem>["itemRenderer"] =>
    (wrapper, item) =>
        wrapper(item.id, onClick);

const defaultProps: GalleryProps<ObjectItem> = {
    hasMoreItems: true,
    isInfiniteLoad: false,
    itemRenderer: itemWrapperFunction({}),
    items: [{ id: "11" as GUID }, { id: "22" as GUID }, { id: "33" as GUID }],
    name: "gallery-test",
    page: 0,
    pageSize: 10,
    paging: false,
    phoneColumns: 2,
    scrollDirection: "vertical",
    style: { container: {} },
    tabletColumns: 3
};

describe("Gallery", () => {
    it("renders correctly", () => {
        const gallery = render(<Gallery {...defaultProps} />);
        expect(gallery).toMatchSnapshot();
    });

    it("renders with empty placeholder", () => {
        const gallery = render(
            <Gallery
                {...defaultProps}
                emptyPlaceholderRenderer={() => <Text>Empty list...</Text>}
                hasMoreItems={false}
                items={[]}
                numberOfItems={0}
            />
        );
        expect(gallery).toMatchSnapshot();
        const emptyPlaceholder = gallery.getByText("Empty list...");
        expect(emptyPlaceholder).toBeDefined();
    });

    it("renders correctly with infinite load ", () => {
        const gallery = render(<Gallery {...defaultProps} isInfiniteLoad />);
        expect(gallery).toMatchSnapshot();
    });

    it("renders correctly with onclick event", () => {
        const gallery = render(
            <Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick: jest.fn() })} />
        );
        expect(gallery).toMatchSnapshot();
    });

    describe("with events", () => {
        it("with on click action", () => {
            const onClick = jest.fn();
            const gallery = render(<Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick })} />);
            const galleryFirstItem = gallery.getByTestId("gallery-test-list-item-22");
            fireEvent.press(galleryFirstItem);
            expect(onClick).toBeCalled();
        });
        it("with pull down action", async () => {
            const pullDown = jest.fn();
            const galleryComponent = render(
                <Gallery {...defaultProps} scrollDirection="vertical" pullDown={pullDown} />
            );
            const galleryFlatList = galleryComponent.getByTestId("gallery-test-list");
            expect(galleryFlatList).toBeDefined();

            const { refreshControl } = galleryFlatList.props;
            await act(async () => {
                refreshControl.props.onRefresh();
            });
            expect(pullDown).toBeCalled();
        });
    });

    describe("with pagination", () => {
        it("renders correctly", () => {
            const gallery = render(<Gallery {...defaultProps} isInfiniteLoad={false} paging hasMoreItems />);
            expect(gallery).toMatchSnapshot();
        });

        it("triggers correct events on click load more items button", () => {
            const loadMoreItems = jest.fn();
            const gallery = render(
                <Gallery
                    {...defaultProps}
                    isInfiniteLoad={false}
                    paging
                    paginationPosition="below"
                    numberOfItems={20}
                    hasMoreItems
                    loadMoreItems={loadMoreItems}
                />
            );
            const loadMoreItemsButton = gallery.getByTestId("gallery-test-pagination-button");
            fireEvent.press(loadMoreItemsButton);
            expect(loadMoreItems).toBeCalled();
        });

        it("triggers correct events on end reached", () => {
            const loadMoreItems = jest.fn();
            const gallery = render(
                <Gallery
                    {...defaultProps}
                    isInfiniteLoad
                    numberOfItems={20}
                    hasMoreItems
                    loadMoreItems={loadMoreItems}
                />
            );
            const galleryList = gallery.getByTestId("gallery-test-list");
            fireEvent.scroll(galleryList, {
                nativeEvent: {
                    contentOffset: {
                        y: 500
                    },
                    contentSize: {
                        height: 500,
                        width: 100
                    },
                    layoutMeasurement: {
                        height: 100,
                        width: 100
                    }
                }
            });
            expect(loadMoreItems).toBeCalled();
        });
    });
});

// describe("Gallery - Tablet", () => {
//     it("renders correctly", () => {
//         const loadMoreItems = jest.fn();

//         const gallery = render(
//             <Gallery
//                 {...defaultProps}
//                 isInfiniteLoad={false}
//                 paging
//                 paginationPosition="below"
//                 numberOfItems={20}
//                 hasMoreItems
//                 loadMoreItems={loadMoreItems}
//             />
//         );
//         expect(gallery).toMatchSnapshot();
//     });
// });
