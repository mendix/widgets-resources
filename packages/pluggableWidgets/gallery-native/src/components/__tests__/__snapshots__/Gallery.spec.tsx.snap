// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery renders correctly 1`] = `
<View
  style={Object {}}
  testID="gallery-test"
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "11",
        },
        Object {
          "id": "22",
        },
        Object {
          "id": "33",
        },
        Object {
          "id": "_blank",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="gallery-test-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-11"
          >
            11
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-22"
          >
            22
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-33"
          >
            33
          </View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 0.5,
                "margin": 0,
                "padding": 0,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery renders correctly with infinite load  1`] = `
<View
  style={Object {}}
  testID="gallery-test"
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "11",
        },
        Object {
          "id": "22",
        },
        Object {
          "id": "33",
        },
        Object {
          "id": "_blank",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="gallery-test-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-11"
          >
            11
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-22"
          >
            22
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-33"
          >
            33
          </View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 0.5,
                "margin": 0,
                "padding": 0,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery renders correctly with onclick event 1`] = `
<View
  style={Object {}}
  testID="gallery-test"
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "11",
        },
        Object {
          "id": "22",
        },
        Object {
          "id": "33",
        },
        Object {
          "id": "_blank",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="gallery-test-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-11"
          >
            11
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-22"
          >
            22
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-33"
          >
            33
          </View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 0.5,
                "margin": 0,
                "padding": 0,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery renders with empty placeholder 1`] = `
<View
  style={Object {}}
  testID="gallery-test"
>
  <RCTScrollView
    ListEmptyComponent={
      <Text>
        Empty list...
      </Text>
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="gallery-test-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <Text
        onLayout={[Function]}
      >
        Empty list...
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Gallery with pagination renders correctly 1`] = `
<View
  style={Object {}}
  testID="gallery-test"
>
  <RCTScrollView
    ListFooterComponent={
      <View>
        <Button
          onPress={[Function]}
          testID="gallery-test-pagination-button"
          title="Load more items..."
        />
      </View>
    }
    data={
      Array [
        Object {
          "id": "11",
        },
        Object {
          "id": "22",
        },
        Object {
          "id": "33",
        },
        Object {
          "id": "_blank",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="gallery-test-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-11"
          >
            11
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-22"
          >
            22
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                undefined,
                Object {
                  "flex": 0.5,
                },
              ]
            }
            testID="gallery-test-list-item-33"
          >
            33
          </View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 0.5,
                "margin": 0,
                "padding": 0,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View>
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="gallery-test-pagination-button"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Load more items...
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
